// sequelize/lib/dialects/abstract/query-generator/dist/transaction.dev.js
"use strict";

var uuidv4 = require("uuid").v4;

var TransactionQueries = {
  setIsolationLevelQuery: function setIsolationLevelQuery(value, options) {
    if (options.parent) {
      return;
    }

    return "SET TRANSACTION ISOLATION LEVEL ".concat(value, ";");
  },
  generateTransactionId: function generateTransactionId() {
    return uuidv4();
  },
  startTransactionQuery: function startTransactionQuery(transaction) {
    if (transaction.parent) {
      return "SAVEPOINT ".concat(this.quoteIdentifier(transaction.name, true), ";");
    }

    return "START TRANSACTION;";
  },
  deferConstraintsQuery: function deferConstraintsQuery() {},
  setConstraintQuery: function setConstraintQuery() {},
  setDeferredQuery: function setDeferredQuery() {},
  setImmediateQuery: function setImmediateQuery() {},
  commitTransactionQuery: function commitTransactionQuery(transaction) {
    if (transaction.parent) {
      return;
    }

    return "COMMIT;";
  },
  rollbackTransactionQuery: function rollbackTransactionQuery(transaction) {
    if (transaction.parent) {
      return "ROLLBACK TO SAVEPOINT ".concat(this.quoteIdentifier(transaction.name, true), ";");
    }

    return "ROLLBACK;";
  }
};
module.exports = TransactionQueries;
