// sequelize/lib/dialects/mariadb/dist/data-types.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var wkx = require("wkx");

var _ = require("lodash");

var momentTz = require("moment-timezone");

var moment = require("moment");

module.exports = function (BaseTypes) {
  BaseTypes.ABSTRACT.prototype.dialectTypes = "https://mariadb.com/kb/en/library/resultset/#field-types";
  BaseTypes.DATE.types.mariadb = ["DATETIME"];
  BaseTypes.STRING.types.mariadb = ["VAR_STRING"];
  BaseTypes.CHAR.types.mariadb = ["STRING"];
  BaseTypes.TEXT.types.mariadb = ["BLOB"];
  BaseTypes.TINYINT.types.mariadb = ["TINY"];
  BaseTypes.SMALLINT.types.mariadb = ["SHORT"];
  BaseTypes.MEDIUMINT.types.mariadb = ["INT24"];
  BaseTypes.INTEGER.types.mariadb = ["LONG"];
  BaseTypes.BIGINT.types.mariadb = ["LONGLONG"];
  BaseTypes.FLOAT.types.mariadb = ["FLOAT"];
  BaseTypes.TIME.types.mariadb = ["TIME"];
  BaseTypes.DATEONLY.types.mariadb = ["DATE"];
  BaseTypes.BOOLEAN.types.mariadb = ["TINY"];
  BaseTypes.BLOB.types.mariadb = ["TINYBLOB", "BLOB", "LONGBLOB"];
  BaseTypes.DECIMAL.types.mariadb = ["NEWDECIMAL"];
  BaseTypes.UUID.types.mariadb = false;
  BaseTypes.ENUM.types.mariadb = false;
  BaseTypes.REAL.types.mariadb = ["DOUBLE"];
  BaseTypes.DOUBLE.types.mariadb = ["DOUBLE"];
  BaseTypes.GEOMETRY.types.mariadb = ["GEOMETRY"];
  BaseTypes.JSON.types.mariadb = ["JSON"];

  var DECIMAL =
  /*#__PURE__*/
  function (_BaseTypes$DECIMAL) {
    _inherits(DECIMAL, _BaseTypes$DECIMAL);

    function DECIMAL() {
      _classCallCheck(this, DECIMAL);

      return _possibleConstructorReturn(this, _getPrototypeOf(DECIMAL).apply(this, arguments));
    }

    _createClass(DECIMAL, [{
      key: "toSql",
      value: function toSql() {
        var definition = _get(_getPrototypeOf(DECIMAL.prototype), "toSql", this).call(this);

        if (this._unsigned) {
          definition += " UNSIGNED";
        }

        if (this._zerofill) {
          definition += " ZEROFILL";
        }

        return definition;
      }
    }]);

    return DECIMAL;
  }(BaseTypes.DECIMAL);

  var DATE =
  /*#__PURE__*/
  function (_BaseTypes$DATE) {
    _inherits(DATE, _BaseTypes$DATE);

    function DATE() {
      _classCallCheck(this, DATE);

      return _possibleConstructorReturn(this, _getPrototypeOf(DATE).apply(this, arguments));
    }

    _createClass(DATE, [{
      key: "toSql",
      value: function toSql() {
        return this._length ? "DATETIME(".concat(this._length, ")") : "DATETIME";
      }
    }, {
      key: "_stringify",
      value: function _stringify(date, options) {
        if (!moment.isMoment(date)) {
          date = this._applyTimezone(date, options);
        }

        return date.format("YYYY-MM-DD HH:mm:ss.SSS");
      }
    }], [{
      key: "parse",
      value: function parse(value, options) {
        value = value.string();

        if (value === null) {
          return value;
        }

        if (momentTz.tz.zone(options.timezone)) {
          value = momentTz.tz(value, options.timezone).toDate();
        } else {
          value = new Date("".concat(value, " ").concat(options.timezone));
        }

        return value;
      }
    }]);

    return DATE;
  }(BaseTypes.DATE);

  var DATEONLY =
  /*#__PURE__*/
  function (_BaseTypes$DATEONLY) {
    _inherits(DATEONLY, _BaseTypes$DATEONLY);

    function DATEONLY() {
      _classCallCheck(this, DATEONLY);

      return _possibleConstructorReturn(this, _getPrototypeOf(DATEONLY).apply(this, arguments));
    }

    _createClass(DATEONLY, null, [{
      key: "parse",
      value: function parse(value) {
        return value.string();
      }
    }]);

    return DATEONLY;
  }(BaseTypes.DATEONLY);

  var UUID =
  /*#__PURE__*/
  function (_BaseTypes$UUID) {
    _inherits(UUID, _BaseTypes$UUID);

    function UUID() {
      _classCallCheck(this, UUID);

      return _possibleConstructorReturn(this, _getPrototypeOf(UUID).apply(this, arguments));
    }

    _createClass(UUID, [{
      key: "toSql",
      value: function toSql() {
        return "CHAR(36) BINARY";
      }
    }]);

    return UUID;
  }(BaseTypes.UUID);

  var GEOMETRY =
  /*#__PURE__*/
  function (_BaseTypes$GEOMETRY) {
    _inherits(GEOMETRY, _BaseTypes$GEOMETRY);

    function GEOMETRY(type, srid) {
      var _this;

      _classCallCheck(this, GEOMETRY);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(GEOMETRY).call(this, type, srid));

      if (_.isEmpty(_this.type)) {
        _this.sqlType = _this.key;
      } else {
        _this.sqlType = _this.type;
      }

      return _this;
    }

    _createClass(GEOMETRY, [{
      key: "toSql",
      value: function toSql() {
        return this.sqlType;
      }
    }], [{
      key: "parse",
      value: function parse(value) {
        value = value.buffer();

        if (!value || value.length === 0) {
          return null;
        }

        value = value.slice(4);
        return wkx.Geometry.parse(value).toGeoJSON({
          shortCrs: true
        });
      }
    }]);

    return GEOMETRY;
  }(BaseTypes.GEOMETRY);

  var ENUM =
  /*#__PURE__*/
  function (_BaseTypes$ENUM) {
    _inherits(ENUM, _BaseTypes$ENUM);

    function ENUM() {
      _classCallCheck(this, ENUM);

      return _possibleConstructorReturn(this, _getPrototypeOf(ENUM).apply(this, arguments));
    }

    _createClass(ENUM, [{
      key: "toSql",
      value: function toSql(options) {
        return "ENUM(".concat(this.values.map(function (value) {
          return options.escape(value);
        }).join(", "), ")");
      }
    }]);

    return ENUM;
  }(BaseTypes.ENUM);

  var JSONTYPE =
  /*#__PURE__*/
  function (_BaseTypes$JSON) {
    _inherits(JSONTYPE, _BaseTypes$JSON);

    function JSONTYPE() {
      _classCallCheck(this, JSONTYPE);

      return _possibleConstructorReturn(this, _getPrototypeOf(JSONTYPE).apply(this, arguments));
    }

    _createClass(JSONTYPE, [{
      key: "_stringify",
      value: function _stringify(value, options) {
        return options.operation === "where" && typeof value === "string" ? value : JSON.stringify(value);
      }
    }]);

    return JSONTYPE;
  }(BaseTypes.JSON);

  return {
    ENUM: ENUM,
    DATE: DATE,
    DATEONLY: DATEONLY,
    UUID: UUID,
    GEOMETRY: GEOMETRY,
    DECIMAL: DECIMAL,
    JSON: JSONTYPE
  };
};
