// sequelize/lib/errors/dist/index.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;

var __markAsModule = function __markAsModule(target) {
  return __defProp(target, "__esModule", {
    value: true
  });
};

var __export = function __export(target, all) {
  __markAsModule(target);

  for (var name in all) {
    __defProp(target, name, {
      get: all[name],
      enumerable: true
    });
  }
};

var __reExport = function __reExport(target, module2, desc) {
  if (module2 && _typeof(module2) === "object" || typeof module2 === "function") {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      var _loop = function _loop() {
        var key = _step.value;
        if (!__hasOwnProp.call(target, key) && key !== "default") __defProp(target, key, {
          get: function get() {
            return module2[key];
          },
          enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable
        });
      };

      for (var _iterator = __getOwnPropNames(module2)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        _loop();
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  return target;
};

var __toModule = function __toModule(module2) {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {
    get: function get() {
      return module2["default"];
    },
    enumerable: true
  } : {
    value: module2,
    enumerable: true
  })), module2);
};

__export(exports, {
  AccessDeniedError: function AccessDeniedError() {
    return import_access_denied_error["default"];
  },
  AggregateError: function AggregateError() {
    return import_aggregate_error["default"];
  },
  AssociationError: function AssociationError() {
    return import_association_error["default"];
  },
  AsyncQueueError: function AsyncQueueError() {
    return import_async_queue.AsyncQueueError;
  },
  BaseError: function BaseError() {
    return import_base_error["default"];
  },
  BulkRecordError: function BulkRecordError() {
    return import_bulk_record_error["default"];
  },
  ConnectionAcquireTimeoutError: function ConnectionAcquireTimeoutError() {
    return import_connection_acquire_timeout_error["default"];
  },
  ConnectionError: function ConnectionError() {
    return import_connection_error["default"];
  },
  ConnectionRefusedError: function ConnectionRefusedError() {
    return import_connection_refused_error["default"];
  },
  ConnectionTimedOutError: function ConnectionTimedOutError() {
    return import_connection_timed_out_error["default"];
  },
  DatabaseError: function DatabaseError() {
    return import_database_error["default"];
  },
  EagerLoadingError: function EagerLoadingError() {
    return import_eager_loading_error["default"];
  },
  EmptyResultError: function EmptyResultError() {
    return import_empty_result_error["default"];
  },
  ExclusionConstraintError: function ExclusionConstraintError() {
    return import_exclusion_constraint_error["default"];
  },
  ForeignKeyConstraintError: function ForeignKeyConstraintError() {
    return import_foreign_key_constraint_error["default"];
  },
  HostNotFoundError: function HostNotFoundError() {
    return import_host_not_found_error["default"];
  },
  HostNotReachableError: function HostNotReachableError() {
    return import_host_not_reachable_error["default"];
  },
  InstanceError: function InstanceError() {
    return import_instance_error["default"];
  },
  InvalidConnectionError: function InvalidConnectionError() {
    return import_invalid_connection_error["default"];
  },
  OptimisticLockError: function OptimisticLockError() {
    return import_optimistic_lock_error["default"];
  },
  QueryError: function QueryError() {
    return import_query_error["default"];
  },
  SequelizeScopeError: function SequelizeScopeError() {
    return import_sequelize_scope_error["default"];
  },
  TimeoutError: function TimeoutError() {
    return import_timeout_error["default"];
  },
  UniqueConstraintError: function UniqueConstraintError() {
    return import_unique_constraint_error["default"];
  },
  UnknownConstraintError: function UnknownConstraintError() {
    return import_unknown_constraint_error["default"];
  },
  ValidationError: function ValidationError() {
    return import_validation_error["default"];
  },
  ValidationErrorItem: function ValidationErrorItem() {
    return import_validation_error.ValidationErrorItem;
  },
  ValidationErrorItemOrigin: function ValidationErrorItemOrigin() {
    return import_validation_error.ValidationErrorItemOrigin;
  },
  ValidationErrorItemType: function ValidationErrorItemType() {
    return import_validation_error.ValidationErrorItemType;
  }
});

var import_base_error = __toModule(require("./base-error"));

var import_database_error = __toModule(require("./database-error"));

var import_aggregate_error = __toModule(require("./aggregate-error"));

var import_association_error = __toModule(require("./association-error"));

var import_bulk_record_error = __toModule(require("./bulk-record-error"));

var import_connection_error = __toModule(require("./connection-error"));

var import_eager_loading_error = __toModule(require("./eager-loading-error"));

var import_empty_result_error = __toModule(require("./empty-result-error"));

var import_instance_error = __toModule(require("./instance-error"));

var import_optimistic_lock_error = __toModule(require("./optimistic-lock-error"));

var import_query_error = __toModule(require("./query-error"));

var import_sequelize_scope_error = __toModule(require("./sequelize-scope-error"));

var import_validation_error = __toModule(require("./validation-error"));

var import_access_denied_error = __toModule(require("./connection/access-denied-error"));

var import_connection_acquire_timeout_error = __toModule(require("./connection/connection-acquire-timeout-error"));

var import_connection_refused_error = __toModule(require("./connection/connection-refused-error"));

var import_connection_timed_out_error = __toModule(require("./connection/connection-timed-out-error"));

var import_host_not_found_error = __toModule(require("./connection/host-not-found-error"));

var import_host_not_reachable_error = __toModule(require("./connection/host-not-reachable-error"));

var import_invalid_connection_error = __toModule(require("./connection/invalid-connection-error"));

var import_exclusion_constraint_error = __toModule(require("./database/exclusion-constraint-error"));

var import_foreign_key_constraint_error = __toModule(require("./database/foreign-key-constraint-error"));

var import_timeout_error = __toModule(require("./database/timeout-error"));

var import_unknown_constraint_error = __toModule(require("./database/unknown-constraint-error"));

var import_unique_constraint_error = __toModule(require("./validation/unique-constraint-error"));

var import_async_queue = __toModule(require("../dialects/mssql/async-queue"));
