// moment/src/lib/duration/dist/get.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.get = get;
exports.weeks = weeks;
exports.years = exports.months = exports.days = exports.hours = exports.minutes = exports.seconds = exports.milliseconds = void 0;

var _aliases = require("../units/aliases");

var _absFloor = _interopRequireDefault(require("../utils/abs-floor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function get(units) {
  units = (0, _aliases.normalizeUnits)(units);
  return this.isValid() ? this[units + 's']() : NaN;
}

function makeGetter(name) {
  return function () {
    return this.isValid() ? this._data[name] : NaN;
  };
}

var milliseconds = makeGetter('milliseconds'),
    seconds = makeGetter('seconds'),
    minutes = makeGetter('minutes'),
    hours = makeGetter('hours'),
    days = makeGetter('days'),
    months = makeGetter('months'),
    years = makeGetter('years');
exports.years = years;
exports.months = months;
exports.days = days;
exports.hours = hours;
exports.minutes = minutes;
exports.seconds = seconds;
exports.milliseconds = milliseconds;

function weeks() {
  return (0, _absFloor["default"])(this.days() / 7);
}
