// sequelize/lib/associations/dist/has-one.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;

var __defNormalProp = function __defNormalProp(obj, key, value) {
  return key in obj ? __defProp(obj, key, {
    enumerable: true,
    configurable: true,
    writable: true,
    value: value
  }) : obj[key] = value;
};

var __spreadValues = function __spreadValues(a, b) {
  for (var prop in b || (b = {})) {
    if (__hasOwnProp.call(b, prop)) __defNormalProp(a, prop, b[prop]);
  }

  if (__getOwnPropSymbols) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = __getOwnPropSymbols(b)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var prop = _step.value;
        if (__propIsEnum.call(b, prop)) __defNormalProp(a, prop, b[prop]);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  return a;
};

var __spreadProps = function __spreadProps(a, b) {
  return __defProps(a, __getOwnPropDescs(b));
};

var Utils = require("./../utils");

var Helpers = require("./helpers");

var _ = require("lodash");

var Association = require("./base");

var Op = require("../operators");

var HasOne =
/*#__PURE__*/
function (_Association) {
  _inherits(HasOne, _Association);

  function HasOne(source, target, options) {
    var _this;

    _classCallCheck(this, HasOne);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(HasOne).call(this, source, target, options));
    _this.associationType = "HasOne";
    _this.isSingleAssociation = true;
    _this.foreignKeyAttribute = {};

    if (_this.as) {
      _this.isAliased = true;
      _this.options.name = {
        singular: _this.as
      };
    } else {
      _this.as = _this.target.options.name.singular;
      _this.options.name = _this.target.options.name;
    }

    if (_.isObject(_this.options.foreignKey)) {
      _this.foreignKeyAttribute = _this.options.foreignKey;
      _this.foreignKey = _this.foreignKeyAttribute.name || _this.foreignKeyAttribute.fieldName;
    } else if (_this.options.foreignKey) {
      _this.foreignKey = _this.options.foreignKey;
    }

    if (!_this.foreignKey) {
      _this.foreignKey = Utils.camelize([Utils.singularize(_this.options.as || _this.source.name), _this.source.primaryKeyAttribute].join("_"));
    }

    if (_this.options.sourceKey && !_this.source.rawAttributes[_this.options.sourceKey]) {
      throw new Error("Unknown attribute \"".concat(_this.options.sourceKey, "\" passed as sourceKey, define this attribute on model \"").concat(_this.source.name, "\" first"));
    }

    _this.sourceKey = _this.sourceKeyAttribute = _this.options.sourceKey || _this.source.primaryKeyAttribute;
    _this.sourceKeyField = _this.source.rawAttributes[_this.sourceKey].field || _this.sourceKey;
    _this.sourceKeyIsPrimary = _this.sourceKey === _this.source.primaryKeyAttribute;
    _this.associationAccessor = _this.as;
    _this.options.useHooks = options.useHooks;

    if (_this.target.rawAttributes[_this.foreignKey]) {
      _this.identifierField = _this.target.rawAttributes[_this.foreignKey].field || _this.foreignKey;
    }

    var singular = _.upperFirst(_this.options.name.singular);

    _this.accessors = {
      get: "get".concat(singular),
      set: "set".concat(singular),
      create: "create".concat(singular)
    };
    return _this;
  }

  _createClass(HasOne, [{
    key: "_injectAttributes",
    value: function _injectAttributes() {
      var newAttributes = _defineProperty({}, this.foreignKey, __spreadValues({
        type: this.options.keyType || this.source.rawAttributes[this.sourceKey].type,
        allowNull: true
      }, this.foreignKeyAttribute));

      if (this.options.constraints !== false) {
        var target = this.target.rawAttributes[this.foreignKey] || newAttributes[this.foreignKey];
        this.options.onDelete = this.options.onDelete || (target.allowNull ? "SET NULL" : "CASCADE");
        this.options.onUpdate = this.options.onUpdate || "CASCADE";
      }

      Helpers.addForeignKeyConstraints(newAttributes[this.foreignKey], this.source, this.target, this.options, this.sourceKeyField);
      Utils.mergeDefaults(this.target.rawAttributes, newAttributes);
      this.target.refreshAttributes();
      this.identifierField = this.target.rawAttributes[this.foreignKey].field || this.foreignKey;
      Helpers.checkNamingCollision(this);
      return this;
    }
  }, {
    key: "mixin",
    value: function mixin(obj) {
      var methods = ["get", "set", "create"];
      Helpers.mixinMethods(this, obj, methods);
    }
  }, {
    key: "get",
    value: function get(instances, options) {
      var _this2 = this;

      var where, Target, instance, results, result, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _instance, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, _instance2;

      return regeneratorRuntime.async(function get$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              where = {};
              Target = this.target;
              options = Utils.cloneDeep(options);

              if (Object.prototype.hasOwnProperty.call(options, "scope")) {
                if (!options.scope) {
                  Target = Target.unscoped();
                } else {
                  Target = Target.scope(options.scope);
                }
              }

              if (Object.prototype.hasOwnProperty.call(options, "schema")) {
                Target = Target.schema(options.schema, options.schemaDelimiter);
              }

              if (!Array.isArray(instances)) {
                instance = instances;
                instances = void 0;
              }

              if (instances) {
                where[this.foreignKey] = _defineProperty({}, Op["in"], instances.map(function (_instance) {
                  return _instance.get(_this2.sourceKey);
                }));
              } else {
                where[this.foreignKey] = instance.get(this.sourceKey);
              }

              if (this.scope) {
                Object.assign(where, this.scope);
              }

              options.where = options.where ? _defineProperty({}, Op.and, [where, options.where]) : where;

              if (!instances) {
                _context.next = 53;
                break;
              }

              _context.next = 12;
              return regeneratorRuntime.awrap(Target.findAll(options));

            case 12:
              results = _context.sent;
              result = {};
              _iteratorNormalCompletion2 = true;
              _didIteratorError2 = false;
              _iteratorError2 = undefined;
              _context.prev = 17;

              for (_iterator2 = instances[Symbol.iterator](); !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                _instance = _step2.value;
                result[_instance.get(this.sourceKey, {
                  raw: true
                })] = null;
              }

              _context.next = 25;
              break;

            case 21:
              _context.prev = 21;
              _context.t0 = _context["catch"](17);
              _didIteratorError2 = true;
              _iteratorError2 = _context.t0;

            case 25:
              _context.prev = 25;
              _context.prev = 26;

              if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
                _iterator2["return"]();
              }

            case 28:
              _context.prev = 28;

              if (!_didIteratorError2) {
                _context.next = 31;
                break;
              }

              throw _iteratorError2;

            case 31:
              return _context.finish(28);

            case 32:
              return _context.finish(25);

            case 33:
              _iteratorNormalCompletion3 = true;
              _didIteratorError3 = false;
              _iteratorError3 = undefined;
              _context.prev = 36;

              for (_iterator3 = results[Symbol.iterator](); !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
                _instance2 = _step3.value;
                result[_instance2.get(this.foreignKey, {
                  raw: true
                })] = _instance2;
              }

              _context.next = 44;
              break;

            case 40:
              _context.prev = 40;
              _context.t1 = _context["catch"](36);
              _didIteratorError3 = true;
              _iteratorError3 = _context.t1;

            case 44:
              _context.prev = 44;
              _context.prev = 45;

              if (!_iteratorNormalCompletion3 && _iterator3["return"] != null) {
                _iterator3["return"]();
              }

            case 47:
              _context.prev = 47;

              if (!_didIteratorError3) {
                _context.next = 50;
                break;
              }

              throw _iteratorError3;

            case 50:
              return _context.finish(47);

            case 51:
              return _context.finish(44);

            case 52:
              return _context.abrupt("return", result);

            case 53:
              return _context.abrupt("return", Target.findOne(options));

            case 54:
            case "end":
              return _context.stop();
          }
        }
      }, null, this, [[17, 21, 25, 33], [26,, 28, 32], [36, 40, 44, 52], [45,, 47, 51]]);
    }
  }, {
    key: "set",
    value: function set(sourceInstance, associatedInstance, options) {
      var oldInstance, alreadyAssociated, tmpInstance;
      return regeneratorRuntime.async(function set$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              options = __spreadProps(__spreadValues({}, options), {
                scope: false
              });
              _context2.next = 3;
              return regeneratorRuntime.awrap(sourceInstance[this.accessors.get](options));

            case 3:
              oldInstance = _context2.sent;
              alreadyAssociated = oldInstance && associatedInstance && this.target.primaryKeyAttributes.every(function (attribute) {
                return oldInstance.get(attribute, {
                  raw: true
                }) === (associatedInstance.get ? associatedInstance.get(attribute, {
                  raw: true
                }) : associatedInstance);
              });

              if (!(oldInstance && !alreadyAssociated)) {
                _context2.next = 9;
                break;
              }

              oldInstance[this.foreignKey] = null;
              _context2.next = 9;
              return regeneratorRuntime.awrap(oldInstance.save(__spreadProps(__spreadValues({}, options), {
                fields: [this.foreignKey],
                allowNull: [this.foreignKey],
                association: true
              })));

            case 9:
              if (!(associatedInstance && !alreadyAssociated)) {
                _context2.next = 14;
                break;
              }

              if (!(associatedInstance instanceof this.target)) {
                tmpInstance = {};
                tmpInstance[this.target.primaryKeyAttribute] = associatedInstance;
                associatedInstance = this.target.build(tmpInstance, {
                  isNewRecord: false
                });
              }

              Object.assign(associatedInstance, this.scope);
              associatedInstance.set(this.foreignKey, sourceInstance.get(this.sourceKeyAttribute));
              return _context2.abrupt("return", associatedInstance.save(options));

            case 14:
              return _context2.abrupt("return", null);

            case 15:
            case "end":
              return _context2.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "create",
    value: function create(sourceInstance, values, options) {
      var _i, _Object$keys, attribute;

      return regeneratorRuntime.async(function create$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              values = values || {};
              options = options || {};

              if (this.scope) {
                for (_i = 0, _Object$keys = Object.keys(this.scope); _i < _Object$keys.length; _i++) {
                  attribute = _Object$keys[_i];
                  values[attribute] = this.scope[attribute];

                  if (options.fields) {
                    options.fields.push(attribute);
                  }
                }
              }

              values[this.foreignKey] = sourceInstance.get(this.sourceKeyAttribute);

              if (options.fields) {
                options.fields.push(this.foreignKey);
              }

              _context3.next = 7;
              return regeneratorRuntime.awrap(this.target.create(values, options));

            case 7:
              return _context3.abrupt("return", _context3.sent);

            case 8:
            case "end":
              return _context3.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "verifyAssociationAlias",
    value: function verifyAssociationAlias(alias) {
      if (typeof alias === "string") {
        return this.as === alias;
      }

      if (alias && alias.singular) {
        return this.as === alias.singular;
      }

      return !this.isAliased;
    }
  }]);

  return HasOne;
}(Association);

module.exports = HasOne;
